src/Microsoft.IdentityModel.JsonWebTokens/JsonWebToken.cs

```diff
@@ -407,115 +461,123 @@ internal void ReadToken(string encodedJson)
-                ReadOnlyMemory<char> hChars = encodedJson.AsMemory(0, Dot1);
+                if (encodedTokenSpan.Slice(Dot4 + 1).IndexOf('.') != -1)
+                    throw LogHelper.LogExceptionMessage(new SecurityTokenMalformedException(LogMessages.IDX14122));
 
-                // header cannot be empty
-                if (hChars.IsEmpty)
+                ReadOnlySpan<char> headerSpan = encodedTokenSpan.Slice(0, Dot1);
+                if (headerSpan.IsEmpty)
                     throw LogHelper.LogExceptionMessage(new ArgumentException(LogMessages.IDX14307));
 
-                byte[] headerAsciiBytes = new byte[hChars.Length];
+                // right number of dots for JWE (4)
+                byte[] headerAsciiBytes = new byte[headerSpan.Length];
 #if NET6_0_OR_GREATER
-                Encoding.ASCII.GetBytes(hChars.Span, headerAsciiBytes);
+                Encoding.ASCII.GetBytes(headerSpan, headerAsciiBytes);
 #else
                 unsafe
                 {
-                    fixed (char* hCharsPtr = hChars.Span)
+                    fixed (char* hCharsPtr = headerSpan)
                     fixed (byte* headerAsciiBytesPtr = headerAsciiBytes)
                     {
-                        Encoding.ASCII.GetBytes(hCharsPtr, hChars.Length, headerAsciiBytesPtr, headerAsciiBytes.Length);
+                        Encoding.ASCII.GetBytes(hCharsPtr, headerSpan.Length, headerAsciiBytesPtr, headerAsciiBytes.Length);
                     }
                 }
 #endif
+
                 HeaderAsciiBytes = headerAsciiBytes;
 
                 try
                 {
-                    Header = CreateHeaderClaimSet(Base64UrlEncoder.UnsafeDecode(hChars));
+                    Header = CreateHeaderClaimSet(Base64UrlEncoder.UnsafeDecode(headerSpan).AsSpan());
                 }
                 catch (Exception ex)
                 {
                     throw LogHelper.LogExceptionMessage(new ArgumentException(LogHelper.FormatInvariant(
                         LogMessages.IDX14102,
-                        LogHelper.MarkAsUnsafeSecurityArtifact(encodedJson.Substring(0, Dot1), t => t.ToString())),
+                        LogHelper.MarkAsUnsafeSecurityArtifact(headerSpan.ToString(), t => t.ToString())),
                         ex));
                 }
 
-                // dir does not have any key bytes
-                ReadOnlyMemory<char> encryptedKeyBytes = encodedJson.AsMemory(Dot1 + 1, Dot2 - Dot1 - 1);
+                // delegating retrieving encrypted Key to the getter on EncryptedKey
+                ReadOnlySpan<char> encryptedKeyBytes = encodedTokenSpan.Slice(Dot1 + 1, Dot2 - Dot1 - 1);
                 if (!encryptedKeyBytes.IsEmpty)
                 {
                     EncryptedKeyBytes = Base64UrlEncoder.UnsafeDecode(encryptedKeyBytes);
-                    _encryptedKey = encodedJson.Substring(Dot1 + 1, Dot2 - Dot1 - 1);
+                    _encryptedKey = encodedTokenSpan.Slice(Dot1 + 1, Dot2 - Dot1 - 1).ToString();
```

Comment:
Can `encryptedKeyBytes` be reused?
```suggestion
                    _encryptedKey = encryptedKeyBytes.ToString();
```